Split simple report generation, introduce `-jsonSimpleReport`

laanwj opened the pull request:
Split generation of simple report into two passes.

- Pass one generates a SimpleReport structure from AnonymityReport
- Pass two generates either JSON or text to console

Add command-line option `-jsonSimpleReport` to specify generation of a JSON simple report.

`-jsonReport` output will also contain the `simpleReport` as one of the fields (I think this was the intention in #35).

Example output:
```json
{
    "hiddenService": "4hwyik7xxwb6pbvb.onion",
    "risks": [
        {
            "severity": "low",
            "title": "Found IP Addresses",
            "description": "",
            "fix": "",
            "items": [
                "1.4.3.2"
            ]
        },
        {
            "severity": "high",
            "title": "Apache mod_status is enabled and accessible",
            "description": "Why this is bad: An attacker can gain very valuable information\nfrom this internal status page including IP addresses, co-hosted services and user activity.",
            "fix": "To fix, disable mod_status or serve it on a different port than the configured hidden service",
            "items": null
        },
        {
            "severity": "medium",
            "title": "Small number of images with EXIF metadata were discovered!",
            "description": "Why this is bad: EXIF metadata can itself deanonymize a user or\nservice operator (e.g. GPS location, Name etc.). Or, when combined, can be used to link anonymous identities together.",
            "fix": "To fix, re-encode all images to strip EXIF and other metadata.",
            "items": [
                "/sample1.jpg"
            ]
        },
        {
            "severity": "critical",
            "title": "Hidden service private key is accessible!",
            "description": "Why this is bad: This can be used to impersonate the service at any point in the future.",
            "fix": "To fix, generate a new hidden service and make sure the private_key file is not reachable from\nthe web root",
            "items": null
        }
    ]
```

Text output remains approximately the same:
```
--------------- OnionScan Report ---------------
Generating Report for: 4hwyik7xxwb6pbvb.onion

Low Risk: Found IP Addresses
         Items Identified:

         1.4.3.2

High Risk: Apache mod_status is enabled and accessible
         Why this is bad: An attacker can gain very valuable information
         from this internal status page including IP addresses, co-hosted services and user activity.
         To fix, disable mod_status or serve it on a different port than the configured hidden service

Medium Risk: Small number of images with EXIF metadata were discovered!
         Why this is bad: EXIF metadata can itself deanonymize a user or
         service operator (e.g. GPS location, Name etc.). Or, when combined, can be used to link anonymous identities together.
         To fix, re-encode all images to strip EXIF and other metadata.
         Items Identified:

         /sample1.jpg

Critical Risk: Hidden service private key is accessible!
         Why this is bad: This can be used to impersonate the service at any point in the future.
         To fix, generate a new hidden service and make sure the private_key file is not reachable from
         the web root
```

s-rah commented on report/simple_report.go:
```diff
@@ -0,0 +1,174 @@
+	var out = NewSimpleReport(report.OnionScanReport.HiddenService)
+
+	if len(report.EmailAddresses) > 0 {
+		out.AddRisk(SEV_LOW, "Found Identities", "", "", report.EmailAddresses)
```
These should be given their own classification rather than SEV_LOW....since they are informational rather than a risk in and of themselves.

s-rah commented on report/simple_report.go:
```diff
@@ -0,0 +1,174 @@
+const SEV_HIGH = "high"
+const SEV_CRITICAL = "critical"
+
+type Risk struct {
```
I love this structure.

s-rah commented on report/simple_report.go:
```diff
@@ -0,0 +1,174 @@
+		buffer.WriteString(risk_levels[risk.Severity] + " " + risk.Title + "\n")
+		if len(risk.Description) > 0 {
+			for _, item := range strings.Split(risk.Description, "\n") {
+				buffer.WriteString("\t " + item + "\n")
```
Since we are here....this might be a good place to do the formatting of the output to console....rather than in the description itself - to ensure we have a consistency....I'm not sure how easy this to do though.

laanwj replied:
Yes, good idea. There's a FormatParagraph function in bitcoin that I could easily port over for this. Could adapt to the actual console width if that information is available but I guess its undesirable to add dependencies on something like ncurses. 
Apparently in golang there is a function [terminal.GetSize](https://godoc.org/golang.org/x/crypto/ssh/terminal#GetSize) in `x/crypto/ssh/terminal` that can be used for this, will give that a try.

s-rah commented on report/simple_report.go:
```diff
@@ -0,0 +1,174 @@
+	}
+
+	if report.FoundApacheModStatus {
+		out.AddRisk(SEV_HIGH, "Apache mod_status is enabled and accessible",
```
For now I think hard coding this is fine but it would be nice to start thinking about this in terms of #73 and how we extend and configure this kind of thing going forward.

laanwj replied:
One thing I could do is to have these as a list of abstract "SimpleReportRule"s which can be plugged in instead of lumping them together in a function.

laanwj commented:
Updated:

- Added SEV_INFO for info-only "risks" (the lists of email addresses, IP addresses, analytics IDs, bitcoin addresses)
- Add a formatting function (and tests) under utils, and use this to format the report to terminal width instead of hard-coding formatting
- Split up `SimpleReport` checks. Instead of having one huge function that does all checks, have a `SimpleReportCheck` interface which is implemented by every check. This is more extensible - plugins can call `RegisterSimpleReportCheck` to register their own checks on top of the standard ones.
- Print a message when no risks were found, instead of just ending the report.

Oh uh. Apparently I broke Travis in the process. I don't understand why: both the tests and gofmt pass locally. It appears some problem with fetching dependencies.

s-rah commented:
This looks great! I think the reason Travis fails is because the actual FormatParagraph function doesn't seem to be in the commit - unless I've missed something.

laanwj commented:
Oh, that'd explain a few things :) Re-trying with actually adding that file.

laanwj commented on report/simple_report.go:
```diff
@@ -0,0 +1,250 @@
+// Standard checks performed for SimpleReport generation
+// Plugins can extend this list by calling RegisterSimpleReportCheck
+var checks = []SimpleReportCheck{
+	&EmailAddressCheck{},
```
Thinking about putting these in a {name, check} map instead, that would allow passing a list of checks to execute in the input (possibly per host).
